File: backend/domain/skill_visibility.py
from __future__ import annotations

import json
from enum import StrEnum
from pathlib import Path
from typing import Literal, cast


SkillVisibilityValue = Literal["private", "team", "enterprise", "public"]
WritableSkillVisibility = Literal["private", "team", "enterprise"]

_SKILL_VISIBILITY_CATALOG = json.loads(
    Path(__file__).with_name("skill-visibilities.json").read_text(encoding="utf-8"),
)

SKILL_VISIBILITY_VALUES = tuple(_SKILL_VISIBILITY_CATALOG["values"])
WRITABLE_SKILL_VISIBILITY_VALUES = tuple(_SKILL_VISIBILITY_CATALOG["writable"])
DEFAULT_SKILL_VISIBILITY = cast(WritableSkillVisibility, _SKILL_VISIBILITY_CATALOG["default"])
SKILL_VISIBILITY_LABELS = {
    visibility: _SKILL_VISIBILITY_CATALOG["labels"][visibility]
    for visibility in SKILL_VISIBILITY_VALUES
}


class SkillVisibility(StrEnum):
    PRIVATE = "private"
    TEAM = "team"
    ENTERPRISE = "enterprise"
    PUBLIC = "public"


if tuple(visibility.value for visibility in SkillVisibility) != SKILL_VISIBILITY_VALUES:
    raise RuntimeError("SkillVisibility enum values must match skill-visibilities.json")

if not set(WRITABLE_SKILL_VISIBILITY_VALUES).issubset(set(SKILL_VISIBILITY_VALUES)):
    raise RuntimeError("Writable skill visibility values must be a subset of skill visibility values")

if DEFAULT_SKILL_VISIBILITY not in WRITABLE_SKILL_VISIBILITY_VALUES:
    raise RuntimeError("Default skill visibility must be writable")


PRIVATE_SKILL_VISIBILITY: SkillVisibilityValue = "private"
TEAM_SKILL_VISIBILITY: SkillVisibilityValue = "team"
ENTERPRISE_SKILL_VISIBILITY: SkillVisibilityValue = "enterprise"
PUBLIC_SKILL_VISIBILITY: SkillVisibilityValue = "public"


def is_skill_visibility(value: str) -> bool:
    return value in SKILL_VISIBILITY_VALUES


def is_writable_skill_visibility(value: str) -> bool:
    return value in WRITABLE_SKILL_VISIBILITY_VALUES


def normalize_skill_visibility(
    value: str | None,
    default: str | None = DEFAULT_SKILL_VISIBILITY,
) -> SkillVisibilityValue:
    normalized = str(value or default or "").strip().lower()
    if not is_skill_visibility(normalized):
        raise ValueError("Invalid visibility")
    return cast(SkillVisibilityValue, normalized)


def normalize_writable_skill_visibility(
    value: str | None,
    default: str | None = DEFAULT_SKILL_VISIBILITY,
) -> WritableSkillVisibility:
    normalized = normalize_skill_visibility(value, default)
    if not is_writable_skill_visibility(normalized):
        raise ValueError("Invalid visibility")
    return cast(WritableSkillVisibility, normalized)

File: frontend/src/lib/skill-visibility.ts
import skillVisibilities from "@/generated/skill-visibilities.json"

const catalog = skillVisibilities as {
  default: string
  values: readonly string[]
  writable: readonly string[]
  labels: Record<string, string>
}

export const SKILL_VISIBILITY_VALUES = catalog.values as readonly ["private", "team", "enterprise", "public"]
export type SkillVisible = (typeof SKILL_VISIBILITY_VALUES)[number]

export const WRITABLE_SKILL_VISIBILITY_VALUES = catalog.writable as readonly ["private", "team", "enterprise"]
export type WritableSkillVisible = (typeof WRITABLE_SKILL_VISIBILITY_VALUES)[number]

export const DEFAULT_SKILL_VISIBILITY = catalog.default as WritableSkillVisible
export const SKILL_VISIBILITY_LABELS = catalog.labels as Record<SkillVisible, string>

export const WRITABLE_SKILL_VISIBILITY_OPTIONS = WRITABLE_SKILL_VISIBILITY_VALUES.map((value) => ({
  value,
  label: SKILL_VISIBILITY_LABELS[value],
}))

export const isSkillVisibility = (value: string): value is SkillVisible =>
  (SKILL_VISIBILITY_VALUES as readonly string[]).includes(value)

export const isWritableSkillVisibility = (value: string): value is WritableSkillVisible =>
  (WRITABLE_SKILL_VISIBILITY_VALUES as readonly string[]).includes(value)

type SkillVisibilityLabels = {
  visibilityPrivate: string
  visibilityTeam: string
  visibilityEnterprise: string
  visibilityPublic?: string
}

export function getSkillVisibilityLabel(visibility: SkillVisible, labels: SkillVisibilityLabels): string {
  if (visibility === "private") {
    return labels.visibilityPrivate
  }
  if (visibility === "team") {
    return labels.visibilityTeam
  }
  if (visibility === "enterprise") {
    return labels.visibilityEnterprise
  }
  return labels.visibilityPublic ?? SKILL_VISIBILITY_LABELS.public
}
